mava/systems/sac/ff_isac.py

```diff
@@ -0,0 +1,598 @@
+    q2_params = q.init(q2_key, obs_single_batched, act_single_batched)
+    q1_target_params = q.init(q1_target_key, obs_single_batched, act_single_batched)
+    q2_target_params = q.init(q2_target_key, obs_single_batched, act_single_batched)
+
+    # Automatic entropy tuning
+    target_entropy = -cfg.system.target_entropy_scale * action_dim
+    target_entropy = jnp.repeat(target_entropy, n_agents).astype(float)
+    # making sure we have shape=(B, A) so broacasting works fine
+    target_entropy = target_entropy[jnp.newaxis, :]
+    if cfg.system.autotune:
+        log_alpha = jnp.zeros_like(target_entropy)
+    else:
+        log_alpha = jnp.log(cfg.system.init_alpha)
+        log_alpha = jnp.broadcast_to(log_alpha, target_entropy.shape)
+
+    # Pack params
+    online_q_params = Qs(q1_params, q2_params)
+    target_q_params = Qs(q1_target_params, q2_target_params)
+    params = SacParams(actor_params, QsAndTarget(online_q_params, target_q_params), log_alpha)
+
+    # Make opt states.
+    actor_opt = optax.adam(cfg.system.policy_lr)
+    actor_opt_state = actor_opt.init(params.actor)
+
+    q_opt = optax.adam(cfg.system.q_lr)
+    q_opt_state = q_opt.init(params.q.online)
+
+    alpha_opt = optax.adam(cfg.system.alpha_lr)
+    alpha_opt_state = alpha_opt.init(params.log_alpha)
+
+    # Pack opt states
+    opt_states = OptStates(actor_opt_state, q_opt_state, alpha_opt_state)
+
+    # Distribute params and opt states across all devices
+    params = replicate(params)
+    opt_states = replicate(opt_states)
+
+    # Create replay buffer
+    init_transition = Transition(
+        obs=obs,
+        action=acts,
+        reward=jnp.zeros((n_agents,), dtype=float),
+        done=jnp.zeros((n_agents,), dtype=bool),
+        next_obs=obs,
+    )
+
+    rb = fbx.make_item_buffer(
+        max_length=cfg.system.buffer_size,
+        min_length=cfg.system.explore_steps,
+        sample_batch_size=cfg.system.batch_size,
+        add_batches=True,
+    )
+    buffer_state = replicate(rb.init(init_transition))
+
+    nns = (actor, q)
+    opts = (actor_opt, q_opt, alpha_opt)
+
+    # Reset env.
+    n_keys = cfg.arch.num_envs * cfg.arch.n_devices * cfg.system.update_batch_size
+    key_shape = (cfg.arch.n_devices, cfg.system.update_batch_size, cfg.arch.num_envs, -1)
+    key, reset_key = jax.random.split(key)
+    reset_keys = jax.random.split(reset_key, n_keys)
+    reset_keys = jnp.reshape(reset_keys, key_shape)
+
+    # Keys passed to learner
+    first_keys = jax.random.split(key, (cfg.arch.n_devices * cfg.system.update_batch_size))
+    first_keys = first_keys.reshape((cfg.arch.n_devices, cfg.system.update_batch_size, -1))
+
+    env_state, first_timestep = jax.pmap(  # devices
+        jax.vmap(  # update_batch_size
+            jax.vmap(env.reset),  # num_envs
+            axis_name="batch",
+        ),
+        axis_name="device",
+    )(reset_keys)
+    first_obs = first_timestep.observation
```

Comment:
Quite a NIT question: 

But in the PPO systems we do this: 
```
    # Initialise environment states and timesteps: across devices and batches.
    key, *env_keys = jax.random.split(
        key, n_devices * config.system.update_batch_size * config.arch.num_envs + 1
    )
    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
        jnp.stack(env_keys),
    )
    reshape_states = lambda x: x.reshape(
        (n_devices, config.system.update_batch_size, config.arch.num_envs) + x.shape[1:]
    )
    # (devices, update batch size, num_envs, ...)
    env_states = jax.tree_map(reshape_states, env_states)
    timesteps = jax.tree_map(reshape_states, timesteps)
```

Is there a reason you are manually `pmap`ing here? Since the train function gets `pmap`ed later on.

Reply:
Ye I pmap here because then all the timespteps are on the correct devices at the start of training, but you could also do it without pmapping, I think it makes it a bit more clear honestly

Reply:
Okay, thanks for clarifying - let's keep it like this in that case then.